// tslint-sonarts/lib/se/stateTransitions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/*
 * SonarTS
 * Copyright (C) 2017-2017 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
const ts = require("typescript");
const tsutils = require("tsutils");
const symbolicValues_1 = require("./symbolicValues");
const tsutils_1 = require("tsutils");
const navigation_1 = require("../utils/navigation");
function applyExecutors(programPoint, state, symbols, shouldTrackSymbol = () => true) {
    const { parent } = programPoint;
    // TODO is there a better way to handle this?
    // special case: `let x;`
    if (parent && tsutils.isVariableDeclaration(parent) && parent.name === programPoint) {
        return variableDeclaration(parent);
    }
    if (tsutils.isNumericLiteral(programPoint)) {
        return numeralLiteral(programPoint);
    }
    if (tsutils.isIdentifier(programPoint)) {
        return identifier(programPoint);
    }
    if (tsutils.isBinaryExpression(programPoint)) {
        return binaryExpression(programPoint);
    }
    if (tsutils.isVariableDeclaration(programPoint)) {
        return variableDeclaration(programPoint);
    }
    if (tsutils.isCallExpression(programPoint)) {
        return callExpression(programPoint);
    }
    if (tsutils.isObjectLiteralExpression(programPoint)) {
        return objectLiteralExpression();
    }
    if (tsutils.isPropertyAccessExpression(programPoint)) {
        return propertyAccessExpression();
    }
    if (tsutils.isPostfixUnaryExpression(programPoint)) {
        return postfixUnaryExpression(programPoint);
    }
    return state.pushSV(symbolicValues_1.simpleSymbolicValue());
    function identifier(identifier) {
        const symbol = symbolAt(identifier);
        let sv = (symbol && state.sv(symbol)) || symbolicValues_1.simpleSymbolicValue();
        return state.pushSV(sv);
    }
    function numeralLiteral(literal) {
        return state.pushSV(symbolicValues_1.numericLiteralSymbolicValue(literal.text));
    }
    function binaryExpression(expression) {
        if (tsutils.isAssignmentKind(expression.operatorToken.kind)) {
            return navigation_1.collectLeftHandIdentifiers(expression.left).identifiers.reduce((state, identifier) => {
                if (expression.operatorToken.kind === ts.SyntaxKind.EqualsToken) {
                    let [value, nextState] = state.popSV();
                    const variable = symbolAt(identifier);
                    if (!variable) {
                        return nextState;
                    }
                    if (!value) {
                        throw new Error("Assignment without value");
                    }
                    return nextState.pushSV(value).setSV(variable, value);
                }
                else {
                    const variable = symbolAt(identifier);
                    const value = symbolicValues_1.simpleSymbolicValue();
                    return variable ? state.pushSV(value).setSV(variable, value) : state;
                }
            }, state);
        }
        return state.pushSV(symbolicValues_1.simpleSymbolicValue());
    }
    function variableDeclaration(declaration) {
        if (tsutils.isIdentifier(declaration.name)) {
            let [value, nextState] = state.popSV();
            const variable = symbolAt(declaration.name);
            if (!variable || !shouldTrackSymbol(variable)) {
                return nextState;
            }
            if (!value) {
                value = symbolicValues_1.undefinedSymbolicValue();
            }
            return nextState.setSV(variable, value);
        }
        return state;
    }
    function callExpression(callExpression) {
        let nextState = state;
        callExpression.arguments.forEach(_ => (nextState = nextState.popSV()[1]));
        nextState = nextState.popSV()[1]; // Pop callee value
        return nextState.pushSV(symbolicValues_1.simpleSymbolicValue());
    }
    function objectLiteralExpression() {
        // TODO it's not so simple. We need to pop plenty of things here
        return state.pushSV(symbolicValues_1.objectLiteralSymbolicValue());
    }
    function propertyAccessExpression() {
        return state.popSV()[1].pushSV(symbolicValues_1.simpleSymbolicValue());
    }
    function postfixUnaryExpression(unary) {
        let nextState = state;
        const operand = unary.operand;
        const sv = symbolicValues_1.simpleSymbolicValue();
        if (tsutils_1.isIdentifier(operand)) {
            const symbol = symbolAt(operand);
            if (symbol) {
                nextState = nextState.setSV(symbol, sv);
            }
        }
        return nextState.popSV()[1].pushSV(sv);
    }
    function symbolAt(node) {
        return (symbols.getUsage(node) || { symbol: null }).symbol;
    }
}
exports.applyExecutors = applyExecutors;
