// tslint-sonarts/lib/rules/noUselessCastRule.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/*
 * SonarTS
 * Copyright (C) 2017-2017 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
const tslint = require("tslint");
const ts = require("typescript");
const navigation_1 = require("../utils/navigation");
class Rule extends tslint.Rules.TypedRule {
    applyWithProgram(sourceFile, program) {
        return this.applyWithWalker(new Walker(sourceFile, this.getOptions(), program));
    }
}
Rule.metadata = {
    ruleName: "no-useless-cast",
    description: "Redundant casts and not-null assertions should be avoided",
    optionsDescription: "",
    options: null,
    rspecKey: "RSPEC-4325",
    type: "functionality",
    typescriptOnly: false,
};
Rule.MESSAGE_CAST = "Remove this unnecessary cast.";
Rule.MESSAGE_ASSERTION = "Remove this unnecessary not-null assertion.";
exports.Rule = Rule;
class Walker extends tslint.ProgramAwareRuleWalker {
    // for some reason there is no specific "visit..." method for AsExpression
    visitNode(node) {
        if (navigation_1.is(node, ts.SyntaxKind.AsExpression, ts.SyntaxKind.TypeAssertionExpression)) {
            this.checkTypeCasting(node);
        }
        super.visitNode(node);
    }
    visitNonNullExpression(node) {
        const typeAtLocation = this.getTypeChecker().getTypeAtLocation(node.expression);
        if (typeAtLocation.getFlags() & ts.TypeFlags.Union) {
            const { types } = typeAtLocation;
            if (!types.some(isUndefinedOrNull)) {
                this.addFailureAtNode(node, Rule.MESSAGE_ASSERTION);
            }
        }
        else if (!isUndefinedOrNull(typeAtLocation)) {
            this.addFailureAtNode(node, Rule.MESSAGE_ASSERTION);
        }
        super.visitNonNullExpression(node);
    }
    checkTypeCasting(assertionExpression) {
        const { expression, type } = assertionExpression;
        const actualExpressionType = this.getTypeChecker().getTypeAtLocation(expression);
        const typeToCast = this.getTypeChecker().getTypeFromTypeNode(type);
        if (actualExpressionType === typeToCast) {
            this.addFailureAtNode(assertionExpression, Rule.MESSAGE_CAST);
        }
    }
}
function isUndefinedOrNull(type) {
    return !!(type.flags & ts.TypeFlags.Undefined) || !!(type.flags & ts.TypeFlags.Null);
}
