// tslint-sonarts/lib/rules/useTypeAliasRule.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/*
 * SonarTS
 * Copyright (C) 2017-2017 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
const tslint = require("tslint");
const ts = require("typescript");
const navigation_1 = require("../utils/navigation");
const sonar_utils_1 = require("../runner/sonar-utils");
class Rule extends tslint.Rules.TypedRule {
    applyWithProgram(sourceFile, program) {
        return this.applyWithWalker(new Walker(sourceFile, this.getOptions(), program));
    }
}
Rule.metadata = {
    ruleName: "use-type-alias",
    description: "Type aliases should be used",
    rationale: tslint.Utils.dedent `
      Union and intersection types are very convenient but can make code a bit harder to read and to maintain. So if a 
      particular union or intersection is used in multiple places it is highly recommended to use a type alias.`,
    optionsDescription: "",
    options: null,
    rspecKey: "RSPEC-4323",
    type: "functionality",
    typescriptOnly: true,
};
Rule.NUMBER_OF_TYPES_THRESHOLD = 3;
Rule.REPEATED_USAGE_THRESHOLD = 3;
exports.Rule = Rule;
class Walker extends tslint.ProgramAwareRuleWalker {
    constructor() {
        super(...arguments);
        this.unionOrIntersectionTypeUsage = new Map();
    }
    visitNode(node) {
        if (navigation_1.is(node, ts.SyntaxKind.UnionType, ts.SyntaxKind.IntersectionType)) {
            const typeNode = node;
            const type = this.getTypeChecker().getTypeFromTypeNode(typeNode);
            // number of types is determined from AST node, because real type will resolve type aliases
            if (isUnionOrIntersectionType(type) && typeNode.types.length >= Rule.NUMBER_OF_TYPES_THRESHOLD) {
                let typeUsages = this.unionOrIntersectionTypeUsage.get(type);
                if (!typeUsages) {
                    typeUsages = [typeNode];
                    this.unionOrIntersectionTypeUsage.set(type, typeUsages);
                }
                else {
                    typeUsages.push(typeNode);
                }
                if (typeUsages.length === Rule.REPEATED_USAGE_THRESHOLD) {
                    const lines = typeUsages.map(u => sonar_utils_1.nodeToSonarLine(u));
                    lines.shift();
                    this.addFailureAtNode(typeUsages[0], Walker.message(type, lines));
                }
            }
        }
        super.visitNode(node);
    }
    visitTypeAliasDeclaration(_) {
        // cut the visit
    }
    static message(type, alsoUsed) {
        const uniqueLines = Array.from(new Set(alsoUsed)).sort((a, b) => a - b);
        const lines = uniqueLines.length > 1 ? ` It is also used on lines ${uniqueLines}.` : "";
        const typeKind = type.flags & ts.TypeFlags.Union ? "union" : "intersection";
        return `Replace this ${typeKind} type with a type alias.${lines}`;
    }
}
function isUnionOrIntersectionType(type) {
    return !!(type.flags & ts.TypeFlags.UnionOrIntersection);
}
